grease/src/Grease/Heuristic.hs

```diff
@@ -425,32 +441,39 @@ pointerHeuristics ::
   MemoryErrorHeuristic sym ext 8 argTys ->
   [RefineHeuristic sym bak ext argTys]
 pointerHeuristics la ptrHeuristic byteHeuristic =
-  [ \bak anns initMem obligation minfo argNames args -> do
-      let sym = C.backendGetSym bak
-      let labeledPred = C.proofGoal obligation
-      let loc = C.simErrorLoc (labeledPred ^. W4.labeledPredMsg)
-      case minfo of
-        Just (_cs, Mem.BBUndefinedBehavior ub) ->
-          handleUB la anns sym loc args ub
-        Just (_cs, Mem.BBMemoryError memErr@(Mem.MemoryError op reason)) -> do
-          let onePtrHeuristics =
-                applyMemoryHeuristics la anns sym ptrHeuristic byteHeuristic loc initMem memErr argNames args
-          case op of
-            Mem.MemLoadOp _ _ ptr _ -> onePtrHeuristics ptr
-            Mem.MemStoreOp _ _ ptr _ -> onePtrHeuristics ptr
-            Mem.MemStoreBytesOp _ ptr _ _ -> onePtrHeuristics ptr
-            Mem.MemLoadHandleOp _ _ ptr _ -> onePtrHeuristics ptr
-            Mem.MemInvalidateOp _ _ ptr _ _ -> onePtrHeuristics ptr
-            Mem.MemCopyOp (_, dst) (_, src) _len _ -> do
-              case reason of
-                Mem.UnreadableRegion -> onePtrHeuristics src
-                Mem.UnwritableRegion -> onePtrHeuristics dst
-                _ -> do
-                  r1 <- onePtrHeuristics dst
-                  r2 <- onePtrHeuristics src
-                  pure (mergeResultsOptimistic r1 r2)
-        Nothing -> pure Unknown
+  [ f
   ]
+ where
+  f bak anns initMem obligation minfo argNames args =
```

Comment:
This is a bit odd, why did this change from a lambda to being in a `where`-clause?

Reply:
this is because we need a type signature on applyToErr to abstract out the two calls to applyMemoryHeuristics, let me know if you have a better way but that ends up in a situation where we now want to bind that applyErr under the lambda with a type sig hence the where clause

Reply:
Can you `let`-bind `applyToErr` with a type signature inwide the `BBMemoryError` case, just before `onePointerHeuristics`?

If not, I'd recommend lifting `f` to a top-level function called `pointerHeuristic` and making `pointerHeuristics` a singleton list around it.